govcd/nsxt_edgegateway_unit_test.go

```diff
@@ -0,0 +1,1781 @@
+										Values: []types.OpenApiIPRangeValues{
+											{
+												StartAddress: "10.10.10.1",
+											},
+										},
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+			want: []netip.Addr{
+				netip.MustParseAddr("10.10.10.1"),
+			},
+			wantErr: false,
+		},
+		{
+			name: "EmptyUplink",
+			args: args{
+				uplinks: []types.EdgeGatewayUplinks{
+					{},
+				},
+			},
+			want:    make([]netip.Addr, 0),
+			wantErr: false,
+		},
+		{
+			name: "EmptySubnets",
+			args: args{
+				uplinks: []types.EdgeGatewayUplinks{
+					{
+						Subnets: types.OpenAPIEdgeGatewaySubnets{},
+					},
+				},
+			},
+			want:    make([]netip.Addr, 0),
+			wantErr: false,
+		},
+		{
+			name: "EmptySubnetValues",
+			args: args{
+				uplinks: []types.EdgeGatewayUplinks{
+					{
+						Subnets: types.OpenAPIEdgeGatewaySubnets{
+							Values: []types.OpenAPIEdgeGatewaySubnetValue{},
+						},
+					},
+				},
+			},
+			want:    make([]netip.Addr, 0),
+			wantErr: false,
+		},
+		{
+			name: "EmptySubnetValueIpRanges",
+			args: args{
+				uplinks: []types.EdgeGatewayUplinks{
+					{
+						Subnets: types.OpenAPIEdgeGatewaySubnets{
+							Values: []types.OpenAPIEdgeGatewaySubnetValue{
+								{
+									IPRanges: &types.OpenApiIPRanges{},
+								},
+							},
+						},
+					},
+				},
+			},
+			want:    make([]netip.Addr, 0),
+			wantErr: false,
+		},
+		{
+			name: "EmptySubnetValueIpRangeValues",
+			args: args{
+				uplinks: []types.EdgeGatewayUplinks{
+					{
+						Subnets: types.OpenAPIEdgeGatewaySubnets{
+							Values: []types.OpenAPIEdgeGatewaySubnetValue{
+								{
+									IPRanges: &types.OpenApiIPRanges{
+										Values: []types.OpenApiIPRangeValues{},
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+			want:    make([]netip.Addr, 0),
+			wantErr: false,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			got, err := flattenEdgeGatewayUplinkToIpSlice(tt.args.uplinks)
+			if (err != nil) != tt.wantErr {
+				t.Errorf("ipSliceFromEdgeGatewayUplinks() error = %v, wantErr %v", err, tt.wantErr)
+				return
+			}
+			if !reflect.DeepEqual(got, tt.want) {
+				t.Errorf("ipSliceFromEdgeGatewayUplinks() = %v, want %v", got, tt.want)
+			}
+		})
+	}
+}
+
+func Test_getUnusedExternalIPAddress(t *testing.T) {
+	type args struct {
+		uplinks         []types.EdgeGatewayUplinks
+		usedIpAddresses []*types.GatewayUsedIpAddress
+		requiredCount   int
+		optionalSubnet  netip.Prefix
+	}
+	tests := []struct {
+		name    string
+		args    args
+		want    []netip.Addr
+		wantErr bool
+	}{
+		{
+			name: "EmptyStructureError",
+			args: args{
+				uplinks:         []types.EdgeGatewayUplinks{},
+				usedIpAddresses: []*types.GatewayUsedIpAddress{{}},
+				requiredCount:   1,
+				optionalSubnet:  netip.Prefix{},
+			},
+			want:    nil,
+			wantErr: true,
+		},
+		{
+			name: "SingleIpAvailable",
+			args: args{
+				uplinks: []types.EdgeGatewayUplinks{
```

Comment:
Just throwing the idea: Would it make sense to have an auxiliary function, that maybe returns a `[]types.EdgeGatewayUplinks` with an `[]types.OpenApiIPRangeValues` input?
It came to my mind seeing all these nested structs which makes the test to suffer a bit of a "hadouken identation" and some duplication of types.

But this is just a thought, in the end one can read it well anyway.

Reply:
I have done an attempt to improve this by adding a helper function `buildSimpleUplinkStructure`. It worked in some cases, but some deeply indented structures remain because if I wrap them under complicated functions it will be even less readable.